text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+RNG types, so making it object safe is desirable.
+
+The methods provided are all optional, implemented in terms of
+`Reader::read`, e.g. as a sketch (no [error handling])
+
+```rust
+fn read_u32(&mut self) -> u32 {
+    let mut buf = [0u8; 4];
+    self.read(&mut buf);
+
+    ((buf[0] as u32) << 24) |
+    ((buf[1] as u32) << 16) |
+    ((buf[2] as u32) << 8) |
+    ((buf[3] as u32) << 0)
+}
+```
+
+They are provided because most RNG types naturally generate one of
+`u32`, `u64`, `f32` or `f64`. Using the `Reader` implementation
+directly to retrieve one of these types (or a value that builds on
+these types) will have to convert from the natural type into bytes and
+then from bytes back to the natural type, and so providing these
+methods (which `Random` implementations can call) removes this
+overhead.
+
+The choice of these 4 types to specialize to is designed to tailor to
+the simulation use-case (many cryptographic uses will be quite happy
+with the bytes interface provided by `Reader`). Those types are the
+most prominent in simulations, and there are generators that produce
+values of those types directly. However, many generators can
+efficiently produce more that this, e.g. the [ChaCha] algorithm can
+very naturally produce 128-bit `i32x4` SIMD vectors (and equally
+naturally produce up to 512-bit AVX512 vectors), and [dSFMT] naturally
+generates `f64x2` vectors. Future algorithms are likely to be designed
+for use with SIMD so it may be desirable to provide some way to
+special case more than just those 4 types.
+
+[ChaCha]: http://en.wikipedia.org/wiki/Salsa20#ChaCha_variant
+[dSFMT]: http://www.math.sci.hiroshima-u.ac.jp/~%20m-mat/MT/SFMT/index.html#dSFMT
+
+These methods are optional, and users are recommended to call
+`RngExt::gen` instead of these, so they are tentatively marked
+`unstable` in case an alternative approach is devised (they could also
+just be stabilised as is, since it is presumably desirable for the
+current four types to be efficient; any other types that we find that
+need this sort overloading can be added as default methods in future).
+
+
+#### Alternatives
+
+Inherit from `Iterator<u32>` instead of `Reader`, replacing `next_u32`
+with a substitute for `read` (the current `Rng` calls it
+`fill_bytes`). Most PRNG algorithms will find it annoying to implement
+`Reader` (they are defined in terms of `u32` or larger, not bytes),
+and, to retain efficiency, there may be some surprising behaviour with
+`Reader`, e.g.
+
+    let mut a = [0, 0];
+    rng.read(&mut a);
+
+    let mut b = [0, 0];
+    rng.read(&mut b[0..1]);
+    rng.read(&mut b[1..2]);
+
+will give different results. That said, this is inherently expected
+with randomness: a true random number generator can not be relied on
+at all.
+
+### The `RngExt` trait
+
+[the `RngExt` trait]: #the-rngext-trait
+
+`RngExt` provides basic useful functionality on RNG. Users will almost always
+want to `use` this trait (not `Rng`) into scope.
```

Comment:
I quite don't understand what you mean; everything listed here will be public in `std::rand`. This comment is referring to the requirement to bring traits into scope to use their methods: `use std::rand::RngExt;`.